packages/peregrine/lib/talons/RootComponents/Product/useProduct.js

```diff
@@ -0,0 +1,94 @@
+ * @param {GraphQLAST}  props.query - The query to fetch a product.
+ * @param {String}      props.urlKey - The url_key of this product.
+ *
+ * @returns {object}    result
+ * @returns {Bool}      result.error - Indicates a network error occurred.
+ * @returns {Bool}      result.loading - Indicates the query is in flight.
+ * @returns {Bool}      result.product - The product's details.
+ */
+export const useProduct = props => {
+    const { cachePrefix, fragment, mapProduct, query, urlKey } = props;
+
+    const [runQuery, { loading, error, data }] = useLazyQuery(query, {
+        // We're manually checking the cache here, so useLazyQuery can skip that.
+        fetchPolicy: 'network-only',
+        variables: {
+            onServer: false,
+            urlKey
+        }
+    });
+
+    const apolloClient = useApolloClient();
+
+    const productFromCache = useMemo(() => {
+        /*
+         * Look up the product in cache first.
+         *
+         * We may have it from a previous query, but we have to manually tell Apollo cache where to find it.
+         * A single product query (as described by https://github.com/magento/graphql-ce/issues/86)
+         * will alleviate the need to do this manual work.
+         *
+         * If the object with the specified `id` is not in the cache, we get `null`.
+         * If the object is in the cache but doesn't have all the fields we need, an error is thrown.
+         *
+         * @see https://www.apollographql.com/docs/react/caching/cache-interaction/#readfragment.
+         */
+        try {
+            return apolloClient.readFragment({
+                id: `${cachePrefix}:${urlKey}`,
+                fragment,
+                variables: { onServer: false }
+            });
+        } catch (e) {
+            // The product is in the cache but it is missing some fields the fragment needs.
+            return null;
+        }
+    }, [apolloClient, cachePrefix, fragment, urlKey]);
+
+    const product = useMemo(() => {
+        if (productFromCache) {
+            return mapProduct(productFromCache);
+        }
+
+        if (data) {
+            const productFromNetwork = data.productDetail.items[0];
+            return mapProduct(productFromNetwork);
+        }
+
+        // The product isn't in the cache and we don't have a response from GraphQL yet.
+        return null;
+    }, [data, mapProduct, productFromCache]);
+
+    useEffect(() => {
+        // Fetch the product from the network.
+        // Note that this always fires, regardless of whether we have a cached product or not:
+        // If we do not have a cached product, we need to go fetch it from the network.
+        // If we do have a cached product, we want to ensure its cache entry doesn't get stale.
+        // In both cases, Apollo will update the cache with the latest data when this returns.
+        runQuery({
+            variables: {
```

Comment:
You already provided these variables to the initial `useLazyQuery` call. Why do we have to provide them again?

Also I think we can just use `useQuery` with `fetchPolicy: 'cache-and-network'` which is exactly the same as what we're doing here except we don't have to manually invoke the query.

Reply:
Great suggestion, looks way better now: https://github.com/magento/pwa-studio/pull/1935/commits/0512dacbe0ef01ed1021f60babe987f7fe64d00d.